src/components/container/InsightMatrix/InsightMatrix.tsx

```diff
@@ -0,0 +1,146 @@
+        if (props.preview) {
+            return createPreviewDataProvider(props.groupByDataConfiguration);
+        }
+        return getDataProvider(props.groupByDataConfiguration);
+    };
+
+    const getComparisonDataProviders = (): InsightMatrixDataProvider<InsightMatrixComparisonDataConfiguration>[] => {
+        if (props.preview) {
+            return props.comparisonDataConfigurations.map(createPreviewDataProvider);
+        }
+        return props.comparisonDataConfigurations.map(getDataProvider);
+    };
+
+    useInitializeView(() => {
+        const groupByDataProvider = getGroupByDataProvider();
+        groupByDataProvider.loadData(startDate, today).then(result => {
+            return { configuration: groupByDataProvider.configuration, result: result };
+        }).then(setGroupByData);
+
+        const comparisonDataProviders = getComparisonDataProviders();
+        Promise.all(comparisonDataProviders.map(comparisonDataProvider => comparisonDataProvider.loadData(startDate, today).then(result => {
+            return { configuration: comparisonDataProvider.configuration, result: result };
+        }))).then(setComparisonData);
+    }, [], [props.preview, props.groupByDataConfiguration, props.comparisonDataConfigurations]);
+
+    if (!groupByData || !comparisonData) {
+        return <div className="mdhui-insight-matrix" ref={props.innerRef}><LoadingIndicator /></div>;
+    }
+
+    const thresholdDaysLookup = computeThresholdDays(startDate, today, groupByData);
+
+    const allThresholds = groupByData.configuration.thresholds.map(threshold => threshold.label).concat(NotEnteredThreshold);
+    const filteredThresholds = allThresholds.filter(threshold => Object.keys(thresholdDaysLookup).includes(threshold));
+
+    const getDefaultDataLabel = (configuration: InsightMatrixDataConfiguration): string => {
+        if (isSurveyDataType(configuration.rawDataType)) {
+            return configuration.rawDataType.stepIdentifier;
+        }
+        const labelKey = getDailyDataTypeDefinition(configuration.rawDataType).labelKey;
+        return labelKey ? language(labelKey) : configuration.rawDataType;
+    };
+
+    return <div className="mdhui-insight-matrix" style={{ gridTemplateColumns: '112px 1fr ' + Array(Math.max(0, comparisonData.length - 1)).fill('1fr').join(' ') }} ref={props.innerRef}>
+        <div className="mdhui-insight-matrix-header-group-by" style={{ background: groupByHeaderBackgroundColor, color: groupByHeaderTextColor }}>
+            {groupByData.configuration.label ?? getDefaultDataLabel(groupByData.configuration)}
+        </div>
+        {comparisonData.length === 0 &&
+            <div className="mdhui-insight-matrix-no-comparison-data">{language('insight-matrix-no-comparison-data')}</div>
+        }
+        {comparisonData.map(data => data.configuration).map((configuration: InsightMatrixComparisonDataConfiguration, index: number) => {
```

Comment:
Can the map/looping here and on line 120 be consolidated into one?

Reply:
It is possible to invert the grid to fill column by column instead of row by row, but the code doesn't really get any more simplified (perhaps less).  You still end up looping something in two places, either the data types to build the header row or the thresholds to build the first column.

I think the row by row rendering is a little easier for me to understand.  The first loop here builds the header row, and then the loop on 120 building each threshold's row.